Refactor index.js to address a few issues

kachkaev opened the pull request:
Closes #3 & #4, possibly addresses #1

kachkaev committed:
Enable async/await in code by adding new babel dependencies

kachkaev committed:
Refactor index.js to address a few issues

Closes #3 & #4, possibly addresses #1

kachkaev commented:
Ouch, just noticed that you did some refactoring yesterday too @jfairbank! 😭  Anyway, feel free to review my changes to decide what we do next.

My PR should close all open issues. The existing test passes, but I haven't added any new ones yet. Will be happy to do so after your code review.

kachkaev commented on index.js:
```diff
@@ -1,109 +1,145 @@
 
-import fs from 'fs';
-import path from 'path';
+import 'babel-polyfill';
```
I know you've already switched to node 8, which supports async/await out of box. However, if we keep this polyfill, we'll be able to support node 6 as well as it's still in maintenance till April 2019.

kachkaev commented on index.js:
```diff
@@ -1,109 +1,145 @@
-import path from 'path';
+import 'babel-polyfill';
+import program from 'commander';
+import { stat, readFile, writeFile, unlink } from 'fs-extra';
```
We can use promisified fs methods from `fs-extra`, so there is no need to use `util` and promisify them manually.

kachkaev commented on index.js:
```diff
@@ -1,109 +1,145 @@
 import sh from 'shelljs';
 
-const filename = (process.argv[2] || '').trim();
+(async () => {
```
I put all the app's logic into a single async function, which make the data flow much easier to read. With sub-functions (which made sense before switching to async/await) the order of steps becomes a bit less obvious and there's either a need for global variables or for passing misc arguments back and forth.

kachkaev commented on index.js:
```diff
@@ -1,109 +1,145 @@
 
-const filename = (process.argv[2] || '').trim();
+(async () => {
+  program
```
`commander` npm module (from where we import `program`) gives us an easy way of defining command line args. My PR only supports one argument, but if you want some data to go down to elm like in v2, that should be easy to arrange.

kachkaev commented on index.js:
```diff
@@ -1,109 +1,145 @@
+
+  // extract key paths
+  const timestamp = +new Date();
+  const mainModule = `RunElmMain${timestamp}`;
```
With a timestamp in the end of the main module name, we can avoid naming collisions and race conditions.

kachkaev commented on index.js:
```diff
@@ -1,109 +1,145 @@
-  if (typeof err === 'object' && 'message' in err) {
-    if (/does not expose `output`/.test(err.message)) {
-      message = `Elm module \`${moduleName}\` does not define a String \`output\` variable.`;
+  } catch (err) {
```
This catch block represents all possible failures and thus standardizes error printing.

kachkaev commented on index.js:
```diff
@@ -1,109 +1,145 @@
+    .version(require(path.resolve(__dirname, '../package.json')).version)
+    .usage('[options] <file>')
+    .option(
+      '--output-var [name]',
```
This option closes #4

kachkaev commented on index.js:
```diff
@@ -1,109 +1,145 @@
+      'output'
+    )
+    .option(
+      '--project-dir [path]',
```
This option closes #3

kachkaev reviewed (commented):
I commented a few bits for you @jfairbank to help you better see the motivation behind my changes. Looking forward to see your review!

kachkaev committed:
Replace babel-polyfill with babel-preset-env

kachkaev committed:
Upgrade eslint and introduce "npm run lint" for command-line linting

kachkaev committed:
Merge @jfairbank's master branch (v2.0.0)

The contents of index.js have not been merged, which means that the args test is failing.
Node 6 support as been (temporary?) removed as in v2.0.0

kachkaev committed:
Remove package-lock.json in favour of yarn.lock + update it

kachkaev committed:
Restore "extra args" feature introduced in v2.0.0

Both tests pass now!

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
+    // eslint-disable-next-line import/no-unresolved
+    .version(require(path.resolve(__dirname, '../package.json')).version)
+    .usage('[options] <file> [arg1 arg2 ...]')
+    .description('Runs a given elm module and prints variable `output : String` to stdout.'
```
Maybe change to:

```
'Runs an Elm file and prints the result of `output` to stdout.'
```

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
+    .version(require(path.resolve(__dirname, '../package.json')).version)
+    .usage('[options] <file> [arg1 arg2 ...]')
+    .description('Runs a given elm module and prints variable `output : String` to stdout.'
+      + ' If `output` is of type `List String -> String`, you can speficy additional command arguments and they will be passed down to elm as `List String`.')
```
```
' `output` can be a constant of type `String` or a function that accepts `List String` and returns `String`. You can supply command line arguments to the Elm file if `output` accepts `List String`.'
```

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
+      + ' If `output` is of type `List String -> String`, you can speficy additional command arguments and they will be passed down to elm as `List String`.')
+    .option(
+      '--output-var [name]',
+      'name of the elm variable to output (should be of type String or List String -> String)',
```
```
'supply a custom constant/function name different from `output` for printing results`
```

kachkaev replied:
I'll make it even shorter: `constant or function name for printing results`. Help will print:

`--output-name [name]  constant or function name for printing results (default: output)`

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
+    )
+    .option(
+      '--project-dir [path]',
+      'specific directory in which to search for elm-package.json or to create it (if different from file location)'
```
```
'specific directory to search for elm-package.json or to create it (if different from Elm file location)'
```

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
 
-  const contents = await compileToString([outputElmFilename], { yes: true });
+    // ensure --output-var is specified adequately
+    if (!outputVarName.match(/^[a-z_]\w*$/)) {
```
Could this validation happen inside the commander config per https://github.com/tj/commander.js/#regular-expression?

jfairbank replied:
Also, we should guard against `main` because the Elm compiler won't accept it if it isn't `Html msg` or `Program ...`

kachkaev replied:
I would not rely much on commander. It's a popular module, but it's not working as expected in quite a few scenarios from my experience. E.g. if we specify a regexp for the option and the given string fails the test, the option becomes equal to true and we have to handle that value manually. In any case, our checks are multi-angled and this cannot be fitted into quite basic set of features that commander provides. I'd keep all the checks next to each other and thus protect the reader from sorting out any extra magic that's done outside `index.js` that they have in front of them.

kachkaev replied:
Good point about `main`! We also have a few other things defined in the main elm module, so I'll add `'init', 'main', 'program', 'sendOutput'` to the list of restricted output names.

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
 
-  return writeFile(outputCompiledFilename, contents);
-}
+    // ensure user moule path is adequate
```
mo**d**ule

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
+    try {
+      const userModuleFileStats = await stat(userModulePath);
+      if (!userModuleFileStats.isFile()) {
+        throw new Error();
```
It might be worth pulling this outside of the try-catch and throwing an explicit error message that it isn't a file.

kachkaev replied:
I'm not sure what for. Current implementation will print you `Error: File '/invalid-path' does not exist`, which looks totally fine to me. I don't see much logic difference between not finding a file,  not finding a project folder, having problems with `output` name or failing to compile the module. They all indicate a problem with the user input and thus can be treated equally. One try / catch for all hiccups that a user may be responsible for.

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
+    try {
+      const projectDirStats = await stat(projectDir);
+      if (!projectDirStats.isDirectory()) {
+        throw new Error();
```
It might be worth pulling this outside of the try-catch and throwing an explicit error message that it isn't a directory.

kachkaev replied:
Currently printing `Error: Provided --project-dir '/invalid-dir' is not a directory`. See my response to your prev comment.

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
+    } catch (err) {
+      throw new Error(`File '${userModulePath}' could not be read`);
+    }
+    const argsRegex = /^output *\w+ *=/;
```
I think I wrote the wrong regex here. We need at least one space. Can you change it to this please?

```js
const argsRegex = /^output +\w+ *=/;
```

kachkaev replied:
Sure! I did not notice that typo too, thanks!

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
+      .some(line => argsRegex.test(line.trim()));
+
+    // read main module template
+    await sh.cd(projectDir);
```
This should be synchronous already, so we don't need `await`.

kachkaev replied:
👌 you're right!

jfairbank commented on index.js:
```diff
@@ -1,130 +1,158 @@
+    // handle error
+    let message;
+    if (typeof err === 'object' && 'message' in err) {
+      if (err.message.indexOf(`does not expose \`${outputVarName}\``) !== -1) {
```
We should probably create our own error types to make inspection cleaner here and also below.

kachkaev replied:
Maybe, but I'm not sure if this is important at the moment. Both to the "special" errors that we handle in a peculiar way are easy to understand ATM. If we add our own error types, the code may look a bit more organised, but will also introduce new abstractions that a reader would have to understand. The simpler the better IMO, we don't want to overengineer.

jfairbank replied:
I actually forgot that these errors are coming from the Elm compiler anyway, so no worries on this one then.

jfairbank reviewed (changes requested):
Great work @kachkaev!

Thanks for all your effort, especially adding more robust validations and error messages.

I like the overall direction, but I'd really like to go back to the helper function approach from before. Definitely keep all the new code, but can you extract out key parts of the logic into helper functions and let the main async IIFE coordinate them instead?

Also, we'll probably want some tests for the new options. Let me know if you have any trouble with that.

Thanks again!

kachkaev commented:
Thanks for your review @jfairbank! I've been working on restoring the new feature you've added in 2.0.0; will finish that soon.

Will be also happy to look at splitting the code back into smaller functions. However, given that there're quite a few variable used in various places, I currently doubt that moving several lines of code to different parts of the file will help others read and maintain the code. To be honest I even see this as a step back, because with these extra functions that are called only once anyway, the eye has to jump between various parts of the file instead of sliding smoothly along the flow of chained tasks. These tasks do not have simple i/o (i.e. each of them takes more than one input and sometimes produces more than one output), so switching to helper functions without introducing nasty global vars is quite challenging.

kachkaev committed:
output var → output name

elm does not have variables and output is either a constant or a function

kachkaev committed:
Do not exit process inside catch to make sure cleanup always completes in "finally"

kachkaev committed:
Re-enable Node 6 and update docs

kachkaev committed:
Address review comments

kachkaev commented:
@jfairbank I updated a few more things that include responses to your review. Also see my comments to your review comments 😃 Thanks for your quick response again! 🎉 

Feel free to check the docs I've added and review everything again. Once we agree about the new implementation, I'm happy to add a few test cases to cover the new features. Guess the new release can be `2.1.0` since we don't break anything.

What benefits do you see in returning to the helper functions? I'd like to understand what perks may outplay the added complexity mentioned in the previous comment.

kachkaev commented:
I'm thinking of adding more test later on this week. What do you think of switching to some testing framework such as [jest](https://facebook.github.io/jest/)? That could make the code a bit more organised and also make it easier to add more test cases in future. Just want to hear your opinion before writing anything.

Here are minor issue I see in `run-tests` now:
* hard to add tests for _expected_ failures
* in order to add tests for `--help` and `--version` options, some restructuring might be required
* execution is synchronous → slower performance when there are more folders with tests
* no easy way to calculate code coverage
* may be hard to read for new maintainers, especially after we add more test cases

jfairbank commented on index.js:
```diff
@@ -1,130 +1,163 @@
+    if (!outputName.match(/^[a-z_]\w*$/)) {
+      throw new Error(`Provided --output-name \`${outputName}\` is not a valid constant or function name in elm.`);
+    }
+    if (['init', 'main', 'program', 'sendOutput'].indexOf(outputName) !== -1) {
```
`Array.prototype.includes` should be available in Node 6.

kachkaev replied:
True: http://node.green/#ES2016-features-Array-prototype-includes

jfairbank commented:
> What benefits do you see in returning to the helper functions? I'd like to understand what perks may outplay the added complexity I mentioned above.

Organization is a large part of it. Each helper function can encapsulate an important step in the flow and the main function can call them. If those functions are well named, then you can still clearly see the steps taking place with less noise by not having the code all in one place. I had planned on eventually refactoring the code into helper modules/functions to wrangle all the "global" variables, but I haven't had time for that. We can leave this as is for now, but this is something I eventually want to tackle.

> I'm thinking of adding more test later on this week. What do you think of switching to some testing framework such as jest? That could make the code a bit more organised and make it easier to add more test cases in future. Just want to hear your opinion before writing anything.

This would be great to use something more robust. The current testing framework was a minimal starting point. A lot of the pain points you bring up could be more easily tested if most of the business logic were in pure functions too, hence why I eventually want a well-factored implementation with helper modules/functions. I'd say let's minimally get the existing tests into Jest and then see about testing "failures". Let's do a separate PR for it too.

Thanks again for you hard work!

jfairbank commented on CHANGELOG.md:
```diff
@@ -1,3 +1,22 @@
+
+### NEW - option `--project-dir`
+
+Adding `--project-dir=/path/to/project` helps elm pick correct `elm-package.json` and `elm-stuff` if the file to run is in a subdirectory of a project.
```
Add `--project-dir=/path/to/project` to help Elm pick the correct `elm-package.json` and `elm-stuff` if the Elm file is in a subdirectory.

jfairbank commented on CHANGELOG.md:
```diff
@@ -1,3 +1,22 @@
+
+### NEW - option `--output-name`
+
+It is possible to specify which constant or function you would like to output by adding `--output-name=myCustomConstantOrFunction`.
```
Specify a custom name for the output constant or function by adding `--output-name=myCustomConstantOrFunction`.

jfairbank commented on CHANGELOG.md:
```diff
@@ -1,3 +1,22 @@
+
+### NEW - meta options
+
+`run-elm --help` prints usage instructions and `run-elm --version` outputs the installed version
```
Add period

jfairbank commented on README.md:
```diff
@@ -74,3 +76,45 @@ Results in this:
+run-elm automatically detects whether `output` is of type `String` or `List String -> String`, so passing no extra arguments to the command will simply make `args` an empty list and will not produce an error.
+Similarly, if `output` is `String`, extra command arguments are simply ignored as redundant.
+
+### Customising output name
```
Customizing Output Name

kachkaev replied:
👌 although my British spell checker does not like this 😃

jfairbank commented on README.md:
```diff
@@ -74,3 +76,45 @@ Results in this:
+
+### Customising output name
+
+It is possible to specify which constant or function you would like to output by adding `--output-name=myCustomConstantOrFunction` to the command:
```
Specify a custom name for the output constant or function by adding `--output-name=myCustomConstantOrFunction` to the command:

jfairbank commented on README.md:
````diff
@@ -74,3 +76,45 @@ Results in this:
+Hello World!
+```
+
+### Running modules in Elm project subdirectories
````
Running Modules in Elm Project Subdirectories

kachkaev replied:
OK. ‘Titleizing’ other headers too.

jfairbank commented on README.md:
```diff
@@ -74,3 +76,45 @@ Results in this:
+
+Typing `run-elm --help` prints usage instructions, which can be handy for recalling what arguments and options the command accepts.
+
+You can also check what version of elm-run you have installed by typing `run-elm --version`.
```
`elm-run` -> `run-elm`

kachkaev commented:
> I had planned on eventually refactoring the code into helper modules/functions to wrangle all the "global" variables

Right. If it comes to separate isolated modules with functions, then I agree that extracting these out of the IIFE would be a good thing. Such functions don't rely on any global variables, can be unit-tested or even turned into their own npm packages. I also agree with you that we can keep things as is for now with an intent to start this kind of refactoring in future releases. Meanwhile, the IIFE with all the ’meat‘ in it is ≈150 LOC, which is manageable, although not ideal.

Good that you agree about switching to a testing framework — I'll be happy to help with that. A quick thought: what if in this PR we simply leave testing more or less as is and just add two new folders to cover new options? Once 2.1.0 is released, we'll have time to approach testing properly while others can start enjoying the new stuff.

And finally thanks for your new comments! Will work on them now 👍

kachkaev committed:
Edit docs based on @jfairbank's review

https://github.com/jfairbank/run-elm/pull/5#pullrequestreview-91539282

kachkaev committed:
Replace Array.prototype.indexOf with Array.prototype.includes

https://github.com/jfairbank/run-elm/pull/5#pullrequestreview-91537391

kachkaev commented:
Done. Will have a quick look at adding two test cases and guess we're done for now.

At what point do you want to replace _next_ in the changelog with 2.1.0 as well as updating the version in package.json? Is this a part of my PR or a commit by you on master just before the release?

kachkaev committed:
Fix how custom output name is handled

kachkaev committed:
Cover new features with tests

kachkaev commented on .gitignore:
```diff
@@ -5,3 +5,4 @@ repl-temp-*
 lib/index.js
 node_modules/
 tests/**/elm-package.json
+!tests/subdirectory/elm-package.json
```
We need to keep `elm-packge.json` for this test because it has a custom value for `"source-directories"`

kachkaev commented on index.js:
```diff
@@ -1,130 +1,163 @@
+    const mainModuleContents = template
+      .replace('{mainModule}', mainModule)
+      .replace('{userModule}', userModule)
+      .replace(/\{output\}/g, outputName);
```
A regexp is now used here because `{output}` needs to be replaced twice in both templates

kachkaev commented on index.js:
```diff
@@ -1,130 +1,163 @@
+    } catch (err) {
+      throw new Error(`File '${userModulePath}' could not be read`);
+    }
+    const argsRegex = new RegExp(`^${outputName} +\\w+ *=`);
```
I completely forgot to replace `/^${outputName} +\\w+ *=/` with a more general regexp and only discovered this after the new test failed 😆

kachkaev commented on tests/args/input.txt:
```diff
@@ -0,0 +1 @@
+Main.elm hello world
```
Because `run-elm` input is a bit more sophisticated now, I switched to a compulsory `input.txt` in each test folder instead of a former optional `args.json`

kachkaev commented on tests/run-tests.js:
```diff
@@ -18,19 +16,17 @@ const testDirs = fs
 
   try {
-    actualOutput = spawnSync(runElmPath, [mainFile, ...args]).stdout;
+    actualOutput = spawnSync(runElmPath, input, { cwd: test.path }).stdout.toString();
```
`cwd` is being set to allow for relative paths in command arguments (e.g. `Main.elm` instead of `/path/to/Main.elm`)

kachkaev commented on tests/run-tests.js:
```diff
@@ -40,12 +36,12 @@ const results = testDirs.map((test) => {
   }
 
-  if (actualOutput.toString() !== expectedOutput.toString()) {
+  if (actualOutput !== expectedOutput) {
```
I changed the type of these two variables from Buffer to String to call `toString()` a bit less often

kachkaev commented on tests/subdirectory/elm-package.json:
```diff
@@ -0,0 +1,15 @@
+    "repository": "https://github.com/user/project.git",
+    "license": "BSD3",
+    "source-directories": [
+        "src"
```
FYI `tests/subdirectory/elm-package.json` has been added to git because of this row

kachkaev reviewed (commented):
🏁

kachkaev commented:
@jfairbank what else would you like me to look at before this PR can be merged?

jfairbank commented:
Looks good @kachkaev. Thanks again for putting this PR together.

I'll take care of updating the release notes in a separate commit.

kachkaev commented:
Awesome! Really looking forward to seeing a new release — just want to share the npm package with colleagues!

I can also find time to look into jest this week after the new version is out.

jfairbank commented:
> I can also find time to look into jest this week after the new version is out.

Sounds great. New release is now out.
